common/argparse.h

```diff
@@ -0,0 +1,893 @@
+// stream's column width when available.
+//
+// ## The `Args` class
+//
+// The `Args` class is not a meaningful type, it serves two purposes: to give a
+// structured name prefix to the inner types, and to allow marking many
+// components `private` and using `friend` to grant access to implementation
+// details.
+//
+// ## Roadmap / TODOs / planned future work
+//
+// - Detect column width when the stream is a terminal and re-flow text.
+// - Implement short help display mode (and enable that by default).
+// - Add help text to one-of values and render it when present.
+// - Add formatting when the stream supports it (colors, bold, etc).
+// - Improve error messages when parsing fails.
+//   - Reliably display the most appropriate usage string.
+//   - Suggest typo corrections?
+//   - Print snippet or otherwise render the failure better?
+// - Add support for responding to shell autocomplete queries.
+// - Finish adding support for setting and printing version information.
+// - Add short option counting support (`-vvv` -> `--verbose=3`).
+//
+class Args {
+  struct Arg;
+  struct Command;
+  class MetaPrinter;
+
+ public:
+  // The result of parsing arguments can be a parse error, a successfully parsed
+  // command line, or a meta-success due to triggering a meta-action during the
+  // parse such as rendering help text.
+  enum class ParseResult {
+    // Signifies an error parsing arguments. It will have been diagnosed using
+    // the streams provided to the parser, and no useful parsed arguments are
+    // available.
+    Error,
+
+    // Signifies that program arguments were successfully parsed and can be
+    // used.
+    Success,
+
+    // Signifies a successful meta operation such as displaying help text
+    // was performed. No parsed arguments are available, and the side-effects
+    // have been directly provided via the streams provided to the parser.
+    MetaSuccess,
+  };
+
+  // Actions are stored in data structures so we use an owning closure to model
+  // them.
+  using ActionT = std::function<void()>;
+
+  // The core argument info used to render help and other descriptive
+  // information. This is used for both options and positional arguments.
+  // Commands use an extended version below.
+  struct ArgInfo {
+    // The name of the argument. For options, this is the long name parsed after
+    // `--`. Option names must also be unique. Conventionally, these are spelled
+    // in lower case with a single dash separating words as that tends to be
+    // especially easy to type and reasonably easy to read.
+    //
+    // For positional arguments this is only used for help text, and is often
+    // spelled with `ALL-CAPS` to make it clear that it is a placeholder name
+    // and does not appear in the actual command line.
+    llvm::StringRef name;
+
+    // Optional short name for options. This must consist of a single character
+    // in the set [a-zA-Z].
+    llvm::StringRef short_name = "";
```

Comment:
If we're going to be using aggregate initialization here, I wonder if `std::optional<char>` would be better? (`.short_name = 'x'` vs `.short_name = "x"`).

Reply:
I don't think my feelings are any stronger than yours, let's leave as-is unless someone else voices an opinion.